src/core/ext/filters/client_channel/lb_policy/oob_backend_metric.cc

```diff
@@ -384,14 +384,23 @@ OrcaWatcher::~OrcaWatcher() {
 }
 
 void OrcaWatcher::SetSubchannel(Subchannel* subchannel) {
+  bool created = false;
   // Check if our producer is already registered with the subchannel.
-  // If not, create a new one, which will register itself with the subchannel.
-  auto* p = static_cast<OrcaProducer*>(
-      subchannel->GetDataProducer(OrcaProducer::Type()));
-  if (p != nullptr) producer_ = p->RefIfNonZero();
-  if (producer_ == nullptr) {
-    producer_ = MakeRefCounted<OrcaProducer>(subchannel->Ref());
-  }
+  // If not, create a new one.
+  subchannel->GetOrAddDataProducer(
+      OrcaProducer::Type(), [&](Subchannel::DataProducerInterface** producer) {
+        if (*producer != nullptr) producer_ = (*producer)->RefIfNonZero();
+        if (producer_ == nullptr) {
+          producer_ = MakeRefCounted<OrcaProducer>();
+          *producer = producer_.get();
+          created = true;
+        }
+      });
+  // If we just created the producer, start it.
+  // This needs to be done outside of the lambda passed to
+  // GetOrAddDataProducer() to avoid deadlocking by re-acquiring the
+  // subchannel lock while already holding it.
+  if (created) producer_->Start(subchannel->Ref());
   // Register ourself with the producer.
   producer_->AddWatcher(this);
```

Comment:
Does the same race exist if the second initializer calls `AddWatcher` on the producer before the first one calls `Start`?

Reply:
I don't think so.  The producer has its own mutex, which it acquires inside of `AddWatcher()`.

Reply:
Are you sure?  If `AddWatcher` is called before `Start`, couldn't you lose a watch?  `MaybeStartStreamLocked` would be a nop in `AddWatcher` and I see nothing in `Start` that looks for existing watches.

Maybe `AddWatcher` needs to optionally start if it hasn't been started yet?

Reply:
I think this would already be handled properly.  Note that `Start()` does not actually start the ORCA stream; what it starts is the connectivity state watch on the subchannel.  When the connectivity state watch reports READY, then if the watcher map is non-empty, we start the ORCA stream:

https://github.com/grpc/grpc/blob/bcd8c991e62b4d6326e6dc61ff1aa3bea2b3f35a/src/core/ext/filters/client_channel/lb_policy/oob_backend_metric.cc#L371

So I don't think it matters if `AddWatcher()` is called before `Start()`.  The only distinction that does matter is whether `AddWatcher()` is called before or after we see READY state, and I think the code is already handling this properly.

Reply:
Ah, I see; thanks for explaining.